File: 백준/Gold/1238. 파티/파티.py
import sys
import math
import heapq

input = sys.stdin.readline
INF = math.inf

n, m, x = map(int, input().split())

arr = [[] for _ in range(n+1)]

for _ in range(m):
    a, b, c = map(int, input().split())

    arr[a].append((b, c))


def dj(start):
    distance = [INF] * (n+1)
    q = []

    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)

        if now == x and start != x:
            return distance

        if dist > distance[now]:
            continue

        for i in arr[now]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))
    return distance


ans = [0 for _ in range(n+1)]

for i in range(1, n+1):
    lst = dj(i)

    # print(arr[i])
    # print(i, lst)
    if x != i:
        ans[i] += lst[x]
    else:
        for idx in range(1, n+1):
            if idx != x:
                ans[idx] += lst[idx]

print(max(ans))

File: 백준/Gold/1238. 파티/파티.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n, m, x;
let dist = [];
const INF = Infinity;
let answer = -1;

rl.on("line", (line) => {
  if (!n) {
    [n, m, x] = line.split(" ").map(Number);
    dist = Array.from({ length: n + 1 }, () =>
      Array.from({ length: n + 1 }, () => INF)
    );
  } else {
    let [a, b, c] = line.split(" ").map(Number);
    dist[a][b] = c;
  }
}).on("close", () => {
  for (let i = 1; i <= n; i++) {
    dist[i][i] = 0;
  }
  floyd();
  for (let i = 1; i <= n; i++) {
    answer = Math.max(dist[i][x] + dist[x][i], answer);
  }
  console.log(answer);
});

function floyd() {
  for (let k = 1; k <= n; k++) {
    for (let i = 1; i <= n; i++) {
      for (let j = 1; j <= n; j++) {
        if (dist[i][k] !== INF && dist[k][j] !== INF) {
          dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
        }
      }
    }
  }
}
